DESIGN.md

```diff
@@ -1,31 +1,379 @@
+
+## Command behaviour
+
+Each command is implemented as an Elixir (or Erlang) module. Command module should
+implement `RabbitMQ.CLI.CommandBehaviour` behaviour.
+
+Behaviour summary:
+
+Following functions MUST be implemented in a command module:
+
+`usage() :: String.t | [String.t]`
+
+Command usage string, or several strings (one per line) to print in command listing in usage.
+Typically looks like `command_name [arg] --option=opt_value`
+
+`flags() :: [Atom.t]`
+
+List of supported option names (e.g. `[:vhost, :timeout]`)
+Error is printed if command doesn't support specified options.
+Global options (except `:vhost`) is ignored.
+Most of the time the keys from the `switches()` callback can be used.
+
+`switches() :: Keyword.t`
+
+Keyword list of switches (argument names and types) for the command.
+
+For example: `def switches(), do: [offline: :boolean, time: :integer]` will
+parse `--offline --time=100` arguments to `%{offline: true, time: 100}` options.
+
+This switches are added to global switches (see [Arguments parsing](#arguments-parsing))
+
+`aliases() :: Keyword.t`
+
+Keyword list of argument names one-letter aliases.
+For example: `[o: :offline, t: :timeout]`
+(see [Arguments parsing](#arguments-parsing))
+
+`banner(arguments :: List.t, options :: Map.t) :: String.t`
+
+Banner to print before the command execution.
+Ignored if argument `--quiet` is specified.
+
+`merge_defaults(arguments :: List.t, options :: Map.t) :: {List.t, Map.t}`
+
+Merge default values for arguments and options (named arguments).
+Returns a tuple with effective arguments and options, that will be passed to `validate/2` and `run/2`
+
+`validate(arguments :: List.t, options :: Map.t) :: :ok | {:validation_failure, Atom.t | {Atom.t, String.t}}`
+
+Validate effective arguments and options.
+
+If function returns `{:validation_failure, err}`
+CLI will print usage to `stderr` and exit with non-zero exit code (typically 64).
+
+`run(arguments :: List.t, options :: Map.t) :: run_result :: any`
+
+Run command. This function usually calls RPC on broker.
+
+`output(run_result :: any, options :: Map.t) :: :ok | {:ok, output :: any} | {:stream, Enum.t} | {:error, ExitCodes.exit_code, [String.t]}`
+
+Cast the return value of `run/2` command to formattable value and exit code.
+
+`:ok` - return `0` exit code and won't print anything
+
+`{:ok, output}` - return `exit` code and print output with `format_output/2` callback in formatter
+
+`{:stream, stream}` - format with `format_stream/2` callback in formatter, iterating over enumerable elements.
+Can return non-zero code, if error occurs during stream processing.
+
+`{:error, exit_code, strings}` - print error messages to `stderr` and return `exit_code` code
+
+There is a default implementation for this callback in `DefaultOutput` module
+
+Most of commands use default implementation via `use RabbitMQ.CLI.DefaultOutput`
```

Comment:
Most of commands use default implementation via => Most of the commands use the default implementation via